First iteration of smart contract upgrade docs (#742)
digital-asset/docs.daml.com: docs/2.10.0/docs/concepts/ledger-model/upgrade-model-in-depth.rst

.. Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

.. _upgrade_model_reference:

The Upgrade Model in Depth - Reference
======================================

This section describes in detail what rules govern package validation
upon upload and how contracts are upgraded or downgraded at runtime.
These topics are for a large part covered in the sections above. This
section acts as a thorough reference.

Static Checks
-------------

Upgrade static checks are performed once alongside other validity checks
when a DAR is uploaded to a participant. DARs deemed invalid for
upgrades are rejected.

DARs upgrade checks are broken down into package-level checks, in turn

> broken down into template and data type-level checks.

Reviewer: Applied

Packages
~~~~~~~~

**Definition:** A *utility package* is a package with no template
definition, no interface definition, and only
non-`serializable <https://github.com/digital-asset/daml/blob/main-2.x/sdk/daml-lf/spec/daml-lf-1.rst#serializable-types>`__
data type definitions. A utility package is typically only made of

> helper functions and constants.

Reviewer: Applied

In the following, packages whose LF version does not support upgrades
(1.15 and earlier) and utility packages are ignored. Everything is as if

> they didn’t exist for the purposes of the package validity check.

Reviewer: Applied

Upon upload to a participant, a DAR is checked against previously
uploaded DARs for upgrade validity. More precisely, for every package

> with name *p* and version *v* present in the uploaded dar:

Reviewer: Applied

1. The participant looks up versions *v_prev* and *v_next* of *p* in its package database, such that *v_prev* is the greatest version of
   *p* smaller than *v*, and *v_next* is the smallest version of *p*
   greater than *v.*

   .. image:: ./images/upgrade-static-checks-order.svg

2. The participant checks that version *v* of *p* is a valid upgrade of
   version *v_prev* of p.

3. The participant checks that version *v_next* of *p* is a valid
   upgrade of version *v* of *p*.

| This means that “being a valid upgrade” for a package is context
  dependent: it depends on what packages are already installed on the
  participant.
| It is also *modular*: checks are performed at the package level. That
  is, a new version of a package is rejected as soon as it contains some
  element which doesn’t properly upgrade its counterpart in the old
  package, even if some other elements do.

Modules
~~~~~~~

The modules of the upgrading package must form a superset of the modules
of the upgraded package. In other words, it is valid to add new modules
but deleting a module leads to a validation error.

> Examples

Author: For the docs team - is it possible to keep these "Examples" as "titles", i.e. slightly bigger and bold, but not show them in the TOC, as the TOC for this page is very polluted with the word "Examples"

Author: Fair enough, @paulbrauner-da could you make these changes?

Reviewer: Sure, will do. Actually Carrie did most of them already.

> Examples

Reviewer: Applied

> ^^^^^^^^

Reviewer: Applied

In the file tree below, package v2 is a potentially valid upgrade of
package v1, assuming ``v2/A.daml`` is a valid upgrade of ``v1/A.daml``.

.. code::

  ├── v1
  │   ├── daml
  │   │   └── A.daml
  │   └── daml.yaml
  └── v2
      ├── daml
      │   ├── A.daml
      │   └── B.daml
      └── daml.yaml

In the file tree below, package v2 cannot possibly be a valid upgrade of
v1 because it doesn’t define module ``B``.

.. code::

  ├── v1
  │   ├── daml
  │   │   ├── A.daml
  │   │   └── B.daml
  │   └── daml.yaml
  └── v2
      ├── daml
      │   └── A.daml
      └── daml.yaml

Templates
~~~~~~~~~~

The templates of the upgrading package must form a superset of the
templates of the upgraded package. In other words, it is valid to add
new templates but deleting a template leads to a validation error.

.. _examples-1:

Examples

> ^^^^^^^^

Reviewer: Applied

Below, the module on the right is a valid upgrade of the module on the
left. But the module on the left is **not** a valid upgrade of the
module on the right because it lacks a definition for template ``T2``.

> .. list-table::

Author: Note to the docs team - we would very much like to be able to show code snippets side by side, which would usually work using the syntax below, and `.. code::` blocks on each element.
However, the formatting that docs.daml.com uses for codeblocks forces them to expand the whole width of the screen.
Is there any way we could change this behavior for the blocks in this document? Otherwise, either they don't have code highlighting, or the doc gets twice as long.

Author: @paulbrauner-da is also taking a look at this

Reviewer: @samuel-williams-da and I finally got to a point where it's reasonably looking. If we can find the time we'll look into improving the margins and getting rid of the drop shadows later.

   :widths: 50 50
   :width: 100%

   * -  module M where

        template T1
          with
            p : Party
          where
            signatory p
     -  module M where

        template T1
          with
            p : Party
          where
            signatory p

        template T2
          with
            p : Party
          where
            signatory p

Template Parameters
~~~~~~~~~~~~~~~~~~~

An upgrading template may add new optional parameters at the end of the
parameter sequence of the upgraded template. The types of the parameters
that the upgrading template has in common with the upgraded template
must be pairwise valid upgrades of the original types.

Adding a parameter in the middle of the parameter sequence leads to a
validation error.

Adding a non-optional parameter at the end of the parameter leads to a
validation error.

.. _examples-2:

Examples

> ^^^^^^^^

Reviewer: Applied

.. Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

.. _upgrade_model_reference:

The Upgrade Model in Depth - Reference
======================================

This section describes in detail what rules govern package validation

> upon upload and how contracts are upgraded or downgraded at runtime.

Reviewer: Thanks. This is handled below in their respective static checks sections but I've made it explicit in this sentence too now. I'll also add it to the runtime semantics section, which I had forgotten to do!

These topics are for a large part covered in the sections above. This
section acts as a thorough reference.

Static Checks
-------------

Upgrade static checks are performed once alongside other validity checks
when a DAR is uploaded to a participant. DARs deemed invalid for
upgrades are rejected.

> DARs upgrade checks are broken down into package-level checks, in turn

Reviewer: Applied suggestion.

broken down into template and data type-level checks.

Packages
~~~~~~~~

> **Definition:** A *utility package* is a package with no template

Author: @paulbrauner-da Could you update this?

> **Definition:** A *utility package* is a package with no template

Reviewer: Good point, done.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  template T
          with
            p : Party

            x1 : Int
          where
            signatory p

     -  template T
          with
            p : Party
          where
            signatory p

Below, the template on the right is **not** a valid upgrade of the
template on the left because it changes the type of ``x1`` from ``Int`` to ``Text``.
``Text`` is not a valid upgrade of ``Int``.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  template T
          with
            p : Party

            x1 : Int
          where
            signatory p
        
     -  template T
          with
            p : Party

            x1 : Text
          where
            signatory p
        
Template Keys
~~~~~~~~~~~~~

An upgrading template cannot modify the key of the upgraded template in
any way.

Adding a key leads to a validation error.

Removing a key leads to a validation error.

Changing the type of a key leads to a validation error.

For 2.9, key types may also not include definitions from outside the current package, or ``daml-stdlib``/``daml-prim``

.. _examples-3:

Examples

> ^^^^^^^^

Reviewer: Applied

   :widths: 50 50
   :width: 100%

   * -  template T
          with
            p : Party

            k : Text
          where
            signatory p

            key (p, k): (Party, Text)

            maintainer (fst key)
        
     -  template T
          with
            p : Party

            k : Text
          where
            signatory p
        
Below, the template on the right is **not** a valid upgrade of the
template on the left because it changes the type of its key.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  template T
          with
            p : Party

            k : Text
          where
            signatory p

            key (p, k): (Party, Text)

            maintainer (fst key)
        
     -  template T
          with
            p : Party

            k : Text
          where
            signatory p

            key (p, 2): (Party, Int)

            maintainer (fst key)

Template Choices
~~~~~~~~~~~~~~~~

The choices of an upgrading template must form a superset of the choices
of the upgraded template. In other words, it is valid to add new choices
but deleting a choice leads to a validation error.

.. _examples-4:

Examples

> ^^^^^^^^

Reviewer: Applied

Below, the template on the right is a valid upgrade of the template on
the left. It adds a choice ``C`` to the previous version of the template.
But the template on the left is **not** a valid upgrade of the template
on the right as it deletes a choice.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  template T
          with
            p : Party
          where
            signatory p

     -  template T
          with
            p : Party
          where
            signatory p

            choice C : ()
              controller p

              do
                return ()

Template Choices - Parameters
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

As with template parameters, an upgrading choice may add new optional
parameters at the end of the parameter sequence of the upgraded choice.
The types of the parameters that the upgrading choice has in common with
the upgraded choice must be pairwise valid upgrades of the original
types.

Adding a parameter in the middle of the parameter sequence leads to a
validation error.

Adding a non-optional parameter at the end of the parameter leads to a
validation error.

Example

> ^^^^^^^

Reviewer: Applied

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  choice C : ()
          with
            x1 : Int
          
          controller p

          do 
            return ()

     -  choice C : ()
          with
            x2 : Optional Text

            x1 : Int

          controller p

          do 
            return ()

Below, the choice on the right is **not** a valid upgrade of the choice
on the left because it drops parameter ``x1``.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  choice C : ()
          with
            x1 : Int

          controller p

          do 
            return ()

     -  choice C : ()
          with

          controller p

          do 
            return ()

Below, the choice on the right is **not** a valid upgrade of the choice
on the left because it changes the type of ``x1`` from ``Int`` to ``Text``. ``Text`` is
not a valid upgrade of ``Int``.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  choice C : ()
          with
            x1 : Int

          controller p

          do 
            return ()

     -  choice C : ()
          with

          controller p

          do 
            return ()

> Template Choices - Return type

Reviewer: Applied

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The return type of an upgrading choice must be a valid upgrade of the
return type of the upgraded choice.

Changing the return type of a choice for a non-valid upgrade leads to a
validation error.

.. _examples-5:

Examples

> ^^^^^^^^

Reviewer: Applied

Below, the choice on the right is **not** a valid upgrade of the choice
on the left because it changes its return type from ``()`` to ``Int``. ``Int`` is
not a valid upgrade of ``()``.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  choice C : ()
          controller p

          do
            return ()

     -  choice C : Int
          controller p

          do
            return 1

Data Types
~~~~~~~~~~

The serializable data types of the upgrading module must form a superset
of the serializable data types of the upgraded package. In other words,
it is valid to add new data types but deleting a data type leads to a
validation error.

Changing the variety of a serializable data type leads to a validation
error. For instance, one cannot change a record type into a variant
type.

> Non-serializable data types are inexistent from the point-of-view of the

Reviewer: Applied

upgrade validity check. Turning a non-serializable data type into a
serializable one amounts to adding a new data type, which is valid.
Turning a serializable data type into a non-serializable one amounts to
deleting this data type, which is invalid.

.. _examples-6:

Examples

> ^^^^^^^^

Reviewer: Applied

on the left because it drops the serializable data type ``A``.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  module M where

        data A = A

     -  module M where

Below, the module on the right is **not** a valid upgrade of the module
on the left because although it adds an optional field to the record
type ``A``, it also turns ``A`` into a non-serializable type, which amounts to
deleting ``A`` from the point of view of the upgrade validity check.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  module M where

        data A = A
          with

     -  module M where

        data A = A 
          with 
            x : Optional (Int -> Int)

Data Types - Records
~~~~~~~~~~~~~~~~~~~~

An upgrading record may add new optional fields at the end of the field
sequence of the upgraded record. The types of the fields that the
upgrading record has in common with the upgraded record must be pairwise
valid upgrades of the original types.

Adding a field in the middle of the field sequence leads to a validation
error.

Adding a non-optional field at the end of the field leads to a
validation error.

.. _examples-7:

Examples

> ^^^^^^^^

Reviewer: Applied

on the left because it adds a new field ``x2`` before ``x1`` instead of adding
it at the end of the field sequence.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  data T = T with
          x1 : Int

     -  data T = T with
          x2 : Optional Text

          x1 : Int

Below, the record on the right is **not** a valid upgrade of the record
on the left because it drops field ``x2``.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  data T = T with
          x1 : Int

          x2 : Text

     -  data T = T with
          x1 : Int

Below, the record on the right is **not** a valid upgrade of the record
on the left because it changes the type of ``x1`` from ``Int`` to ``Text``. ``Text`` is
not a valid upgrade of ``Int``.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  data T = T with
          x1 : Int

     -  data T = T with
          x1 : Text

Data Types - Variants
~~~~~~~~~~~~~~~~~~~~~

An upgrading variant may add new constructors at the end of the
constructor sequence of the upgraded variant. The types of the arguments
of the constructors that the upgrading variant has in common with the

> upgraded variant must be pairwise valid upgrades of the original types.

Reviewer: Applied

Adding a constructor in the middle of the constructor sequence leads to
a validation error.

Changing the order of the constructor sequence leads to a validation
error.

Removing a constructor leads to a validation error.

.. _examples-8:

Examples

> ^^^^^^^^

Reviewer: Applied

arguments.

Data Types - Type References
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A type reference is an identifier that resolves to a type. For instance,
consider the following module definitions, from two different packages:

.. code:: daml

  -- In package q
  module Dep where

  data U = U with x : Int
  type A = U

.. code:: daml

  -- In package p
  module M where
  import qualified Dep

  data T = T with x : Dep.A

In the definition of ``T``, ``Dep.A`` is a type reference that resolves to the
type with qualified name ``Dep.U`` in package ``q``.

A reference *r2* to a data type upgrades a reference *r1* to a data type
if and only if:

-  *r2* resolves to a type *t2* with qualified name *q2* in package *p2;*

-  *r1* resolves to a type *t1* with qualified name *q1* in package *p1;*

-  The qualified names *q2* and *q1* are the same;

-  If the LF versions of both *p1* and *p2* support upgrades then
      package *p2* is a valid upgrade of package *p1*, else *p2* and
      *p1* are the exact same package.

It is worth noting that even when *t2* upgrades *t1*, *r2* only upgrades
*r1* provided that package *p2* is a valid upgrade of package *p1* as a
whole.

.. _examples-9:

Examples

> ^^^^^^^^

Reviewer: Applied

which is not a valid upgrade of package ``p-2.0.0``, even though the two
definitions of ``V`` are the same.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  module Main where

        -- imported from q-2.0.0

        import qualified Dep

        data T = T Dep.V

     -  module Main where

        -- imported from q-1.0.0

        import qualified Dep

        data T = T Dep.V

Suppose now that q-1.0.0 and q-2.0.0 are both compiled to LF version
1.15 (which does not support upgrades). Then below, the module on the
right is **not** a valid upgrade of the module on the left because the
references to U on each side resolve to packages with different IDs.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  module Main where

        -- imported from q-1.0.0

        import qualified Dep

        data T = T Dep.U

     -  module Main where

        -- imported from q-2.0.0

        import qualified Dep

data T = T Dep.U

Data Types - Parameterized Data Types
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Parameterized data types are considered serializable for the purpose of
upgrade validation. That is, parameterized data types in an upgrading

> package will be compared against their previous version.

Reviewer: Applied

The upgrade validation for parameterized data types follows the same
rules as non-parameterized data types, but also compares type variables.
The type variable comparison is currently limited to a pure syntactic
equality check, meaning that a type variable only upgrades a type
variable of the same name.

.. _example-1:

Example

> ^^^^^^^

Reviewer: Applied

Below, the parameterized data type on the right is a valid upgrade of
the parameterized data type on the left. As is valid with any record
type, it adds an optional field.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  data Tree a = 
          Tree with 
            label : a

            children : [Tree a]

     -  data Tree a = 
          Tree with 
            label : a

            children : [Tree a]

            cachedSize : Optional Int

Below, the parameterized data type on the right is **not** a valid
upgrade of the parameterized data type on the left. It renames the type
variable ``a`` into ``b`` which is currently not supported.

.. list-table::
   :widths: 50 50
   :width: 100%

   * -  data Tree b = 
          Tree with 
            label : b

            children : [Tree b]

     -  data Tree b = 
          Tree with 
            label : b

            children : [Tree b]

Interfaces
~~~~~~~~~~

> Interfaces can’t be upgraded. As a result we strongly discourage

Reviewer: Applied

uploading a package that contains both interface and template
definitions, as these templates won’t be able to benefit from ZDT

> upgrades in the future. Instead, we recommend declaring interfaces in a

Reviewer: Applied

package of their own that defines no template.

Data Transformation: Runtime Semantics
--------------------------------------

Whenever a contract is fetched or one of its choices is exercised, it is
retrieved from the ledger and transformed into a value that fits the
target template type. Below, we detail the rules governing target types
and then explain how transformations are performed.

Target Types
~~~~~~~~~~~~

In a top level fetch or exercise triggered by a Ledger API command, the

> target template type is determined by the rules of package preference

Reviewer: Applied

detailed in the `Ledger API <#_swcg5hoxdia8>`__ section. Once a given
version of a template has been selected, the target type of its
parameters as well as the target type of the contracts retrieved in its
choices is determined by their static type.

.. _example-1-1:

Example 1

> ^^^^^^^^^

Reviewer: Applied

Assume a package ``p`` with two versions. The upgrading versions adds an

> optional text field

Reviewer: Applied

| ID         |               |                                         |
+============+===============+=========================================+
| ``1234``   | ``p-1.0.0:T`` | ``T { p = ‘Alice’ }``                   |
+------------+---------------+-----------------------------------------+
| ``5678``   | ``p-2.0.0:T`` | ``T { p = ‘Bob’, t = Some “Hello” }``   |
+------------+---------------+-----------------------------------------+

Then

-  Fetching contract ``1234`` with package preference ``p-1.0.0`` retrieves the
      contract and leaves it unchanged, returning ``T { p = ‘Alice’ }``.

-  Fetching contract ``1234`` with package preference ``p-2.0.0`` retrieves the
      contract and successfully transforms it to the target template
      type, returning ``T { p = ‘Alice’, t = None }``.

-  Fetching contract ``5678`` with package preference ``p-1.0.0`` retrieves the
      contract and fails to downgrade it to the target template type,
      returning an error.

-  Fetching contract ``5678`` with package preference ``p-2.0.0`` retrieves the
      contract and leaves it unchanged, returning ``T { p = ‘Bob’, t =
      Some “Hello” }``.

Example 2

> ^^^^^^^^^

Reviewer: Applied

and a contract of type ``U`` written by ``dep-2.0.0``.

+-------------+------------------+------------------------------------+
| Contract ID | Type             | Contract                           |
+=============+==================+====================================+
| ``4321``    | ``q:T``          | ``T { p = ‘Alice’ }``              |
+-------------+------------------+------------------------------------+
| ``8765``    | ``dep-2.0.0:U``  | ``U { p = ‘Bob’, t = None }``      |
+-------------+------------------+------------------------------------+

When exercising choice ``GetU 8765`` on contract ``4321`` with package
preference ``dep-2.0.0``, we trigger a fetch of contract ``5678``. Because
package ``q`` depends on version ``1.0.0`` of ``dep``, the target type for ``U`` is the
one defined in package ``dep-1.0.0``. Contract ``5678`` is thus downgraded to ``U
{ p = ‘Bob’}`` upon retrieval. Note that the command preference for
version ``2.0.0`` of package ``dep`` bears no incidence here.

Transformation Rules
~~~~~~~~~~~~~~~~~~~~

Once the target type has been determined, the data transformation rules
themselves are pretty straightforward and follow the `upgrading rules of

> protocol

Reviewer: Applied

.. Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
.. SPDX-License-Identifier: Apache-2.0

.. _upgrade_model_reference:

The Upgrade Model in Depth - Reference
======================================

This section describes in detail what rules govern package validation upon

> upload and how contracts, choice arguements and choice results are upgraded or

Reviewer 2: ```suggestion
upload and how contracts, choice arguments and choice results are upgraded or
```

downgraded at runtime. These topics are for a large part covered in the
sections above. This section acts as a thorough reference.

Static Checks
-------------

Upgrade static checks are performed once alongside other validity checks
when a DAR is uploaded to a participant. DARs deemed invalid for
upgrades are rejected.

DAR upgrade checks are broken down into package-level checks, which are in turn

> broken down into template and data type-level checks.

Reviewer 2: ```suggestion
broken down into module, template and data type-level checks.
```

Packages
~~~~~~~~

**Definition:** A *utility package* is a package with no template
definition, no interface definition, and only
non-`serializable <https://github.com/digital-asset/daml/blob/main-2.x/sdk/daml-lf/spec/daml-lf-1.rst#serializable-types>`__
data type definitions. A utility package typically consists of

> helper functions and constants.

Reviewer 2: ```suggestion
helper functions and constants, but no type definitions.
```

In the following validity check, packages whose LF version does not support upgrades
(1.15 and earlier) and utility packages are ignored. 

A DAR is checked against previously
uploaded DARs for upgrade validity on upload to a participant. Specifically, for every package

> with name *p* and version *v* present in the uploaded DAR:

Reviewer 2: add a ref to package name and version

1. The participant looks up versions *v_prev* and *v_next* of *p* in its package database, such that *v_prev* is the greatest version of
   *p* smaller than *v*, and *v_next* is the smallest version of *p*
   greater than *v.*

   .. image:: ./images/upgrade-static-checks-order.svg

> 2. The participant checks that version *v* of *p* is a valid upgrade of

Reviewer 2: For the sake of clarity, maybe we should say that if no v_prev exists the check is skip.

   version *v_prev* of p.
3. The participant checks that version *v_next* of *p* is a valid
   upgrade of version *v* of *p*.

This means that "being a valid upgrade" for a package is context

> dependent: it depends on what packages are already installed on the

Reviewer 2: ```suggestion
dependent: it depends on what packages are already uploaded on the
```

participant.
It is also *modular*: checks are performed at the package level. That
is, a new version of a package is rejected as soon as it contains some
element which doesn't properly upgrade its counterpart in the old
package, even if some other elements do.

> 

Reviewer 2: maybe we should add if a set of packages is submitted and one of them fail the check, everything is rejected.

**Examples**

Below, the module on the right is a valid upgrade of the module on the
left. But the module on the left is **not** a valid upgrade of the
module on the right because it lacks a definition for template ``T2``.

.. list-table::
   :widths: 50 50
   :width: 100%

   * - .. code-block:: daml

          module M where
          
          template T1      
            with           
              p : Party    
            where          
              signatory p  
    
     - .. code-block:: daml

          module M where   
          
            template T1
              with
                p : Party
              where
                signatory p
          
            template T2
              with
                p : Party
              where
                signatory p
    
Template Parameters
~~~~~~~~~~~~~~~~~~~

An upgrading template may add new optional parameters at the end of the
parameter sequence of the upgraded template. The types of the parameters
that the upgrading template has in common with the upgraded template
must be pairwise valid upgrades of the original types.

Adding a parameter in the middle of the parameter sequence leads to a

> validation error.

Reviewer 2: One cannot change the name of the parameters.

   :width: 100%

   * - .. code-block:: daml

            template T
              with
                p : Party
                x1 : Int
              where
                signatory p

     - .. code-block:: daml

            template T
              with
                p : Party
              where
                signatory p

Below, the template on the right is **not** a valid upgrade of the
template on the left because it changes the type of ``x1`` from ``Int`` to ``Text``.
``Text`` is not a valid upgrade of ``Int``.

.. list-table::
   :widths: 50 50
   :width: 100%

   * - .. code-block:: daml

            template T
              with
                p : Party
                x1 : Int
              where
                signatory p

     - .. code-block:: daml

            template T
              with
                p : Party
                x1 : Text
              where
                signatory p
        
Template Keys
~~~~~~~~~~~~~

An upgrading template cannot modify the key of the upgraded template in
any way.

Adding a key leads to a validation error.

Removing a key leads to a validation error.

Changing the type of a key leads to a validation error.

> For 2.9, key types can only use definitions from the current package or 
> from ``daml-stdlib``/``daml-prim``.

Reviewer 2: ```suggestion
For 2.9, key types can only use definitions from the current package or 
from the Daml standard library. 
```
You could add a  ref to ``daml-stdlib``/``daml-prim``, but you need a link. It is maybe not obvious what those are.

but deleting a choice leads to a validation error.

.. _examples-4:

**Examples**

Below, the template on the right is a valid upgrade of the template on
the left. It adds a choice ``C`` to the previous version of the template.
But the template on the left is **not** a valid upgrade of the template
on the right as it deletes a choice.

.. list-table::
   :widths: 50 50
   :width: 100%

   * - .. code-block:: daml

            template T
              with
                p : Party
              where
                signatory p

     - .. code-block:: daml

            template T
              with
                p : Party
              where
                signatory p

                choice C : ()
                  controller p
                  do
                    return ()

Template Choices - Parameters
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

As with template parameters, an upgrading choice may add new optional
parameters at the end of the parameter sequence of the upgraded choice.
The types of the parameters that the upgrading choice has in common with
the upgraded choice must be pairwise valid upgrades of the original

> types.

Reviewer 2: It is not valid to change the name of a parameters

Below, the module on the right is **not** a valid upgrade of the module
on the left because it drops the serializable data type ``A``.

.. list-table::
   :widths: 50 50
   :width: 100%

   * - .. code-block:: daml

            module M where
     
            data A = A
     
     - .. code-block:: daml

            module M where
     
Below, the module on the right is **not** a valid upgrade of the module
on the left because although it adds an optional field to the record
type ``A``, it also turns ``A`` into a non-serializable type, which amounts to
deleting ``A`` from the point of view of the upgrade validity check.

.. list-table::
   :widths: 50 50
   :width: 100%

   * - .. code-block:: daml

            module M where

            data A = A
              with

     - .. code-block:: daml

            module M where

            data A = A 
              with 
                x : Optional (Int -> Int)

Data Types - Records
~~~~~~~~~~~~~~~~~~~~

An upgrading record may add new optional fields at the end of the field
sequence of the upgraded record. The types of the fields that the
upgrading record has in common with the upgraded record must be pairwise

> valid upgrades of the original types.

Reviewer 2: the name of the fields cannot change.

            x1 : Int
  
Below, the record on the right is **not** a valid upgrade of the record
on the left because it drops field ``x2``.

.. list-table::
   :widths: 50 50
   :width: 100%

   * - .. code-block:: daml

           data T = T with
             x1 : Int
             x2 : Text
     
     - .. code-block:: daml

           data T = T with
             x1 : Int

Below, the record on the right is **not** a valid upgrade of the record
on the left because it changes the type of ``x1`` from ``Int`` to ``Text``. 
``Text`` is not a valid upgrade of ``Int``.

.. list-table::
   :widths: 50 50
   :width: 100%

   * - .. code-block:: daml

           data T = T with
             x1 : Int

     - .. code-block:: daml

           data T = T with
             x1 : Text

Data Types - Variants
~~~~~~~~~~~~~~~~~~~~~

An upgrading variant may add new constructors at the end of the
constructor sequence of the upgraded variant. The argument types  
of the constructors that the upgrading variant has in common with the  
upgraded variant must be pairwise valid upgrades of the original types.  

Adding a constructor in the middle of the constructor sequence leads to
a validation error.

> Changing the order of the constructor sequence leads to a validation
> error.

Reviewer 2: ```suggestion
Changing the order or the name of the constructor sequence leads to a validation
error.
```

which is not a valid upgrade of package ``p-2.0.0``, even though the two
definitions of ``V`` are the same.

.. list-table::
   :widths: 50 50
   :width: 100%

   * - .. code-block:: daml

            module Main where
     
            -- imported from q-2.0.0
            import qualified Dep
     
            data T = T Dep.V

     - .. code-block:: daml

            module Main where
     
            -- imported from q-1.0.0
            import qualified Dep
     
            data T = T Dep.V

Suppose now that q-1.0.0 and q-2.0.0 are both compiled to LF version
1.15 (which does not support upgrades). Then below, the module on the
right is **not** a valid upgrade of the module on the left because the
references to U on each side resolve to packages with different IDs.

.. list-table::
   :widths: 50 50
   :width: 100%

   * - .. code-block:: daml

            module Main where
     
            -- imported from q-1.0.0
            import qualified Dep
     
            data T = T Dep.U
     
     - .. code-block:: daml

            module Main where
     
            -- imported from q-2.0.0
            import qualified Dep
     
            data T = T Dep.U

> Data Types - Parameterized Data Types

Reviewer 2: We are missing non-parametric builtin types.

Reviewer: Added

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Parameterized data types are considered serializable for the purpose of

Reviewer 2: Found this statement a bit strange. This is not only for the purpose of upgrade validation, but for all purpose.

upgrade of the parameterized data type on the left. It renames the type
variable ``a`` into ``b`` which is currently not supported.

.. list-table::
   :widths: 50 50
   :width: 100%

   * - .. code-block:: daml

            data Tree b = 
              Tree with 
                label : b
                children : [Tree b]

     - .. code-block:: daml

            data Tree b = 
              Tree with 
                label : b
                children : [Tree b]

Interfaces
~~~~~~~~~~

Interfaces cannot be upgraded. We strongly discourage
uploading a package that contains both interface and template
definitions, as these templates cannot benefit from smart contract upgrades
in the future. Instead, we recommend declaring interfaces in a
package of their own that defines no template.

Data Transformation: Runtime Semantics
--------------------------------------

Whenever a contract is fetched or one of its choices is exercised, it is
retrieved from the ledger and transformed into a value that fits the target
template type. Then, its metadata (signatories, observers, key, maintainers) is
recomputed using the code of the target temmpate and compared against the

> existing metadata: it is not allowed to change.

Reviewer 2: ensure clause is also reevaluate.

Reviewer: Added .

   :widths: 50 50
   :width: 100%

   * - .. code-block:: daml

            data V =
              = C1 T1
              | ...
              | Cn Tn

     - .. code-block:: daml

            data V =
              = C1 T1'
              | ...
              | Cn Tn'
              | D1 U1
              | ...
              | Dm Um

-  A ``V-v1`` value ``Ci vi`` is upgraded to a ``V-v2`` value by upgrading ``vi``
   recursively. The transformation results in a value ``Ci vi'`` where
   ``vi'`` is the result of upgrading ``vi`` to ``Ti'``.
-  A ``V-v2`` value ``Ci vi`` is downgraded to a ``V-v1`` value by downgrading ``vi``
   recursively. The transformation results in a value ``Ci vi'`` where
   ``vi'`` is the result of downgrading ``vi`` to ``Ti``.
-  Attempting to downgrade a ``V-v2`` value of the form ``Dj vj`` results in a
   runtime error.

The same transformation rules apply to enum types, constructor arguments
aside.

Other Types
^^^^^^^^^^^

Types that aren't records or variants are "pass-through" for the upgrade
and downgrade transformations:

-  Values of scalar types are trivially transformed to themselves.
-  The elements of Lists are recursively transformed.

> -  The keys and values of Maps are recursively transformed.

Reviewer 2: we are missing optional

Reviewer: Added an example too.

                p : Party
                x1 : Int
              where
                signatory p

     - .. code-block:: daml

            template T
              with
                p : Party
              where
                signatory p

Below, the template on the right is **not** a valid upgrade of the
template on the left because it changes the type of ``x1`` from ``Int`` to ``Text``.
``Text`` is not a valid upgrade of ``Int``.

.. list-table::
   :widths: 50 50
   :width: 100%
   :class: diff-block

   * - .. code-block:: daml

            template T
              with
                p : Party
                x1 : Int
              where
                signatory p

     - .. code-block:: daml

            template T
              with
                p : Party
                x1 : Text
              where
                signatory p
        
Template Keys
~~~~~~~~~~~~~

An upgrading template cannot modify the key of the upgraded template in
any way.

Adding a key leads to a validation error.

Removing a key leads to a validation error.

Changing the type of a key leads to a validation error.

For 2.9, key types can only use definitions from the current package or 

> from the Daml standard library. 

Reviewer: That's fine because all types in the std lib are either stable or non-serializable.

            choice C : ()
              with
                x1 : Int
              controller p
              do 
                return ()

     - .. code-block:: daml

            choice C : ()
              with
              controller p
              do 
                return ()

Template Choices - Return Type
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The return type of an upgrading choice must be a valid upgrade of the
return type of the upgraded choice.

Changing the return type of a choice for a non-valid upgrade leads to a
validation error.

.. _examples-5:

**Examples**

Below, the choice on the right is **not** a valid upgrade of the choice
on the left because it changes its return type from ``()`` to ``Int``. ``Int`` is
not a valid upgrade of ``()``.

.. list-table::
   :widths: 50 50
   :width: 100%
   :class: diff-block

   * - .. code-block:: daml

            choice C : ()
              controller p
              do
                return ()

     - .. code-block:: daml

            choice C : Int
              controller p
              do
                return 1

Data Types
~~~~~~~~~~

The serializable data types of the upgrading module must form a superset

> of the serializable data types of the upgraded package. In other words,

Reviewer 3: Technical question: how do I know whether I'm declaring a serializable datatype or not?

Reviewer: We want to have a better story for that for 2.10/3.1: a way to annotate types for serializability (or non-serializability maybe). But right now you have to rely on the LF spec to know whether your type is serializable or not, and you can't force it to be non-serializable.

            x1 : Int

     - .. code-block:: daml

          data T = T with
            x2 : Optional Text
            x1 : Int
  
Below, the record on the right is **not** a valid upgrade of the record
on the left because it drops field ``x2``.

.. list-table::
   :widths: 50 50
   :width: 100%
   :class: diff-block

   * - .. code-block:: daml

           data T = T with
             x1 : Int
             x2 : Text
     
     - .. code-block:: daml

           data T = T with
             x1 : Int

Below, the record on the right is **not** a valid upgrade of the record
on the left because it changes the type of ``x1`` from ``Int`` to ``Text``. 
``Text`` is not a valid upgrade of ``Int``.

.. list-table::
   :widths: 50 50
   :width: 100%
   :class: diff-block

   * - .. code-block:: daml

           data T = T with
             x1 : Int

     - .. code-block:: daml

           data T = T with
             x1 : Text

Data Types - Variants
~~~~~~~~~~~~~~~~~~~~~

An upgrading variant may add new constructors at the end of the
constructor sequence of the upgraded variant. The argument types  
of the constructors that the upgrading variant has in common with the  

> upgraded variant must be pairwise valid upgrades of the original types.  

Reviewer 3: Not clarified here: 
1. Can I add another field to a constructor at the end of its argument sequence? 
2. What happens if I want to add a field to a nullary constructor?
3. What considerations are there wrt avoiding to convert enums into variants?

Reviewer: Addressed all three points and added examples.

            data T =
              A Int | B Text

     - .. code-block:: haskell

            data T = 
              A Int | B Bool

Data Types - Enums
~~~~~~~~~~~~~~~~~~

For the purpose of upgrade validation, enums can be treated as a special
case of variants. The rules of `the section on
variants <#data-types---variants>`__ apply, only without constructor
arguments.

Data Types - Type References
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A type reference is an identifier that resolves to a type. For instance,
consider the following module definitions, from two different packages:

.. code:: daml

  -- In package q
  module Dep where

  data U = U with x : Int
  type A = U

.. code:: daml

  -- In package p
  module M where
  import qualified Dep

  data T = T with x : Dep.A

In the definition of ``T``, ``Dep.A`` is a type reference that resolves to the
type with qualified name ``Dep.U`` in package ``q``.

A reference *r2* to a data type upgrades a reference *r1* to a data type
if and only if:

-  *r2* resolves to a type *t2* with qualified name *q2* in package *p2;*
-  *r1* resolves to a type *t1* with qualified name *q1* in package *p1;*
-  The qualified names *q2* and *q1* are the same;

> -  If the LF versions of both *p1* and *p2* support upgrades then

Reviewer 3: ```suggestion
-  The LF versions of both *p1* and *p2* support upgrades and
```

Reviewer 3: At least that's how I expect the check to work, as otherwise LF 1.15 types could be upgrades of each other.

Reviewer: I meant to say:
```
if (supports_upgrades(p1.version) && supports_upgrades(p2.version))
  then p2 `upgrades` p1
  else p1.pkgid == p2.pkgid
```
The 1.15 case is covered by the else branch.

I feel like this is what the sentence is saying. Maybe the "else" isn't obvious?

Reviewer: Changed for an either...or to make it clearer.

.. list-table::
   :widths: 50 50
   :width: 100%
   :class: diff-block

   * - .. code-block:: daml

            module Main where
     
            -- imported from q-2.0.0
            import qualified Dep
     
            data T = T Dep.V

     - .. code-block:: daml

            module Main where
     
            -- imported from q-1.0.0
            import qualified Dep
     
            data T = T Dep.V

Suppose now that q-1.0.0 and q-2.0.0 are both compiled to LF version
1.15 (which does not support upgrades). Then below, the module on the
right is **not** a valid upgrade of the module on the left because the
references to U on each side resolve to packages with different IDs.

.. list-table::
   :widths: 50 50
   :width: 100%
   :class: diff-block

   * - .. code-block:: daml

            module Main where
     
            -- imported from q-1.0.0
            import qualified Dep
     
            data T = T Dep.U
     
     - .. code-block:: daml

            module Main where
     
            -- imported from q-2.0.0
            import qualified Dep
     
            data T = T Dep.U

Data Types - Parameterized Data Types
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Parameterized data types are considered serializable. That is, parameterized

Reviewer 3: > Parameterized data types are considered serializable. 

This does not hold in this generality, does it? The assumption is that the parameters are considered serializable in the context of the serializability check.

Reviewer 3: It looks like qualifying the sentence with "For the purpose of upgrade validation, parameterized data types are considered serializable, and thus checked for compatibility."

Reviewer: I had this exact same sentence but Remy suggested I removed the "for the purpose of" part. This is because it actually holds generally: we do consider parameterized types serializable. But it's kind of a technicality because you only care about the serializability of these type constructors when fully applied in practice. 

@remyhaemmerle-da what do you think?

Reviewer: As discussed, I just removed that paragraph. I have also since them added alpha equivalence so I reworked the examples.

.. _example-1:

**Example**

Below, the parameterized data type on the right is a valid upgrade of
the parameterized data type on the left. As is valid with any record
type, it adds an optional field.

.. list-table::
   :widths: 50 50
   :width: 100%
   :class: diff-block

   * - .. code-block:: daml

            data Tree a = 
              Tree with 
                label : a
                children : [Tree a]

     - .. code-block:: daml

            data Tree a = 
              Tree with 
                label : a
                children : [Tree a]
                cachedSize : Optional Int

Below, the parameterized data type on the right is **not** a valid
upgrade of the parameterized data type on the left. It renames the type
variable ``a`` into ``b`` which is currently not supported.

.. list-table::
   :widths: 50 50
   :width: 100%
   :class: diff-block

   * - .. code-block:: daml

            data Tree b = 
              Tree with 
                label : b
                children : [Tree b]

     - .. code-block:: daml

            data Tree b = 
              Tree with 
                label : b
                children : [Tree b]

Interfaces
~~~~~~~~~~

> Interfaces cannot be upgraded. We strongly discourage

Reviewer 3: ```suggestion
Interfaces definitions cannot be upgraded. We strongly discourage
```

Reviewer 3: missing: interface instances

uploading a package that contains both interface and template
definitions, as these templates cannot benefit from smart contract upgrades
in the future. Instead, we recommend declaring interfaces in a
package of their own that defines no template.

Data Transformation: Runtime Semantics
--------------------------------------

Whenever a contract is fetched or one of its choices is exercised, it is
retrieved from the ledger and transformed into a value that fits the target
template type. Then, its metadata (signatories, observers, key, maintainers) is
recomputed using the code of the target temmpate and compared against the
existing metadata: it is not allowed to change. The ensure clause of the 

> contract is also re-evaluated.

Reviewer 3: ```suggestion
contract is also re-evaluated: it must evaluate to ``True``.
```

In addition, when a choice is exercised, its arguments are transformed into
values that fit the type of the parameters of the choice in the target package.
The result of the exercise is then possibly transformed back to some other
target type by the client (e.g. the generated java client code).

Below, we detail the rules governing target types, then explain how
transformations are performed, and finally detail the rules of metadata
re-computation.

Target Types
~~~~~~~~~~~~

In a top-level fetch or exercise triggered by a Ledger API command, the
target template type is determined by the rules of package preference
detailed in the `Ledger API <#_swcg5hoxdia8>`__ section. Once a given
version of a template has been selected, the target type of its
parameters as well as the target type of the contracts retrieved in its

> choices is determined by their static type.

Reviewer 3: Unclear at his point what you mean by "their static type".

Reviewer: I mean that the type is determined at compile time. I'll try to find a better way to formulate it.

Reviewer: Clarified.

.. _example-1-1:

**Example 1**

> Assume a package ``p`` with two versions. The upgrading version adds an

Reviewer 3: ```suggestion
Assume a package ``p`` with two versions. The upgraded version adds an
```

Reviewer: In our internal conversations at least, we refer to the old package as "the upgraded" one and the new package as the "upgrading" one. Looks like this terminology is not obvious? What do you think?

Reviewer: Changed everywhere for "the new version of X" and "the prior version of X".

Changing the type of a key leads to a validation error.

> For 2.9, key types may also not include definitions from outside the current package, or ``daml-stdlib``/``daml-prim``

Author: @dylant-da @remyhaemmerle-da @paulbrauner-da let me know what you think.

Reviewer: I find the "or" a bit hard to parse because of the negation. What about the opposite?

"For 2.9, key types can only use definitions from the current package or ``daml-stdlib``/``daml-prim``"

Reviewer: Done in upcoming commit.
